### PR TITLE
[feat] - Chunk servers heartbeat monitoring background task

### PR DESCRIPTION
**Pull request by micheal-o**
Background task to monitor the heartbeat of registered chunkservers from the master server.

Changes:
1. Implementation for ChunkServerHeartBeatMonitorTask and some config manager changes
2. test
3. Added some log messages to chunkserver manager.

### CODE REVIEW

File: src/server/master_server/run_master_server_main.cc [hunk 1]
```diff
@@ -72,9 +74,18 @@ int main(int argc, char** argv) {
 
   LOG(INFO) << "Server listening on " << server_address;
 
+  // Start the chunk server heartbeat monitor task after services have been
+  // started.
+  auto chunk_servers_heartbeat_task =
+      ChunkServerHeartBeatMonitorTask::GetInstance();
+  chunk_servers_heartbeat_task->Start(config_path);
```
Comment (xicheng87) [inline src/server/master_server/run_master_server_main.cc:81]: Can we directly pass the config here as the ConfigManager is already constructed?

File: src/server/master_server/chunk_server_heartbeat_monitor_task.cc [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+ChunkServerHeartBeatMonitorTask*
+ChunkServerHeartBeatMonitorTask::GetInstance() {
+  static ChunkServerHeartBeatMonitorTask instance;
+
+  return &instance;
+}
+
+void ChunkServerHeartBeatMonitorTask::Start(const std::string& config_file) {
```
Comment (xicheng87) [inline src/server/master_server/chunk_server_heartbeat_monitor_task.cc:39]: I think we should just pass the ConfigManager* here as we always construct this in the main, see comments below. I have extended the other service using similar pattern.

File: src/server/master_server/chunk_server_heartbeat_monitor_task.cc [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+         ChunkServerManager::GetInstance().chunk_servers_map_) {
+      // Check if we have been asked to terminate before proceeding.
+      if (IsTerminateSignalled()) {
+        return;
+      }
+
+      CheckHeartBeatRequest request;
+
+      auto server_address =
+          absl::StrCat(this->config_mgr_->ResolveHostname(
```
Comment (xicheng87) [inline src/server/master_server/chunk_server_heartbeat_monitor_task.cc:84]: Should resolve host name be only relevant when resolve_hostname is true? In this vein, I think the constructor param needs to take resolve_hostname in the main function. I have encountered similar issue in chunk server manager's service and fixed with passing (config, resolve_hostname) and only resolve hosname when resolve_hostname is true.

**Comment by Michael-Tu** [comment]
Note @xicheng87 @micheal-o 
Each test, if spinning up a separate test server in the background, should have listen on a **separate**  port; Otherwise, when bazel runs all tests, there will be test files trying to listen on the same port, and it fails

**Comment by xicheng87** [comment]
Gan is correct, and I took a quick look and I think the use of port 50051 is the culprit in this case.

**Comment by Michael-Tu** [comment]
> Gan is correct, and I took a quick look and I think the use of port 50051 is the culprit in this case.

can you fix it?

**Comment by micheal-o** [comment]
> 
> 
> > Gan is correct, and I took a quick look and I think the use of port 50051 is the culprit in this case.
> 
> can you fix it?



> 
> 
> > Gan is correct, and I took a quick look and I think the use of port 50051 is the culprit in this case.
> 
> can you fix it?

fixed the test config to use different ports

**Comment by xicheng87** [comment]
Hi Macheal, were you able to reproduce the failure shown in the build on your side? Just want to make sure that the port is the only issue but nothing else. I tried to change the port in a checkout of this branch but still see the test failed. If you are fairly sure about this, feel free to merge it up.

**Comment by micheal-o** [comment]
> 
> 
> Hi Macheal, were you able to reproduce the failure shown in the build on your side? Just want to make sure that the port is the only issue but nothing else. I tried to change the port in a checkout of this branch but still see the test failed. If you are fairly sure about this, feel free to merge it up.

The test never failed locally even at the time I created this PR. I think the test may be failing here due to parallel runs (guessing), causing several tests to use the same port at the same time. So changed the  port and will see if it fails again. Can't reproduce locally, always passing when I do bazel test ...

File: src/server/master_server/run_master_server_main.cc (hunk 1 above)
Comment (micheal-o) [reply src/server/master_server/run_master_server_main.cc]: done. You're correct. Was just using the existing pattern and was being lazy to change it. I'm getting weak man, can't wait to get this done...lol.

File: src/server/master_server/chunk_server_heartbeat_monitor_task.cc (hunk 2 above)
Comment (micheal-o) [reply src/server/master_server/chunk_server_heartbeat_monitor_task.cc]: ditto

File: src/server/master_server/chunk_server_heartbeat_monitor_task.cc (hunk 3 above)
Comment (micheal-o) [reply src/server/master_server/chunk_server_heartbeat_monitor_task.cc]: now passing resolve_hostname

**Comment by xicheng87** [comment]
Hmm, I suspect that there is something else going on here. Would you mind trying building and testing it with debug mode? Do `bazel test -c dbg ...` . This may take a little while as it would recompile all things with debug symbol.

**Comment by micheal-o** [comment]
Screenshot from 2020-06-04 22-38-44

**Pull request merged**